Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -156,59 +211,49 @@ std::string DecodeString(pb_istream_t* stream) {
 // TODO(rsgowman): Refactor to use a helper class that wraps the stream struct.
 // This will help with error handling, and should eliminate the issue of two
 // 'EncodeFieldValue' methods.
-void EncodeFieldValueImpl(pb_ostream_t* stream, const FieldValue& field_value) {
+void EncodeFieldValueImpl(pb_ostream_t* raw_stream,
+                          const FieldValue& field_value) {
   // TODO(rsgowman): some refactoring is in order... but will wait until after a
   // non-varint, non-fixed-size (i.e. string) type is present before doing so.
+  PbOstream stream(raw_stream);
   bool status = false;
   switch (field_value.type()) {
     case FieldValue::Type::Null:
-      status = pb_encode_tag(stream, PB_WT_VARINT,
-                             google_firestore_v1beta1_Value_null_value_tag);
-      if (!status) {
-        // TODO(rsgowman): figure out error handling
-        abort();
-      }
-      EncodeNull(stream);
+      stream.EncodeTag(PB_WT_VARINT,
+                       google_firestore_v1beta1_Value_null_value_tag);
+      stream.EncodeNull();
       break;
 
     case FieldValue::Type::Boolean:
-      status = pb_encode_tag(stream, PB_WT_VARINT,
-                             google_firestore_v1beta1_Value_boolean_value_tag);
-      if (!status) {
-        // TODO(rsgowman): figure out error handling
-        abort();
-      }
-      EncodeBool(stream, field_value.boolean_value());
+      stream.EncodeTag(PB_WT_VARINT,
+                       google_firestore_v1beta1_Value_boolean_value_tag);
+      stream.EncodeBool(field_value.boolean_value());
       break;
 
     case FieldValue::Type::Integer:
-      status = pb_encode_tag(stream, PB_WT_VARINT,
-                             google_firestore_v1beta1_Value_integer_value_tag);
-      if (!status) {
-        // TODO(rsgowman): figure out error handling
-        abort();
-      }
-      EncodeInteger(stream, field_value.integer_value());
+      stream.EncodeTag(PB_WT_VARINT,
```

Comment:
Can tag encoding be moved into the `Encode*` methods? Or is it realistically possible for a value encoded with `EncodeVarint` to be tagged something other than `PB_WT_VARINT`?

Reply:
It could. (In fact, it used to be.) But we pulled it back out due to an asymmetry with the decode methods.

tl;dr for the decode is that while EncodeVarint/Bool/Int all use PB_WT_VARINT (so mapping is easy) but the reversing it is tricky since it's difficult to know if a PB_WT_VARINT should be decoded as a varint/bool/int. (EncodeString/Submessage also have the same problem as they both map to PB_WT_STRING.)

This is something we can take another look at, but it's probably easiest to do that after the rest of the refactoring (incl decoding) is done.

Reply:
Agreed, let's revisit further on.